app/models/concerns/smart_proxy_host_extensions.rb

```diff
@@ -0,0 +1,52 @@
+module SmartProxyHostExtensions
+  extend ActiveSupport::Concern
+
+  module ClassMethods
+    def smart_proxy_reference(hash)
```

Comment:
This and other methods would deserve some documentation on what options they expect and how they achieve the stuff, since they use quite a lot of reflections and other dark magic, it's hard to read and understand the code (which might be a reason why I might be unconsciously avoiding reviewing the PR :)